File: agenttel-node/src/anomaly/detector.ts
import type { AnomalyResult } from '../interfaces.js';
import { normalResult } from '../interfaces.js';

/** Z-score-based anomaly detector. */
export class AnomalyDetector {
  private readonly zScoreThreshold: number;

  constructor(zScoreThreshold = 3.0) {
    this.zScoreThreshold = zScoreThreshold > 0 ? zScoreThreshold : 3.0;
  }

  /** Evaluates whether the current value is anomalous given baseline stats. */
  evaluate(metric: string, current: number, baselineMean: number, baselineStddev: number): AnomalyResult {
    if (baselineStddev <= 0) return normalResult();

    const zScore = (current - baselineMean) / baselineStddev;
    const absZ = Math.abs(zScore);
    const isAnomaly = absZ >= this.zScoreThreshold;

    let anomalyScore = 0;
    if (isAnomaly) {
      anomalyScore = Math.min(1.0, absZ / this.zScoreThreshold / 2.0);
    }

    return { anomalyScore, isAnomaly, zScore };
  }
}

File: agenttel-go/anomaly/detector.go
// Package anomaly provides z-score anomaly detection and incident pattern matching.
package anomaly

import (
	"math"

	"go.agenttel.dev/agenttel-go/models"
)

// Detector performs z-score-based anomaly detection.
type Detector struct {
	zScoreThreshold float64
}

// NewDetector creates an anomaly detector with the given z-score threshold.
func NewDetector(zScoreThreshold float64) *Detector {
	if zScoreThreshold <= 0 {
		zScoreThreshold = 3.0
	}
	return &Detector{zScoreThreshold: zScoreThreshold}
}

// Evaluate determines if the current value is anomalous given baseline stats.
func (d *Detector) Evaluate(metric string, current, baselineMean, baselineStddev float64) models.AnomalyResult {
	if baselineStddev <= 0 {
		return models.Normal()
	}

	zScore := (current - baselineMean) / baselineStddev
	absZ := math.Abs(zScore)
	isAnomaly := absZ >= d.zScoreThreshold

	// Normalize anomaly score to 0-1 range
	anomalyScore := 0.0
	if isAnomaly {
		anomalyScore = math.Min(1.0, absZ/d.zScoreThreshold/2.0)
	}

	return models.AnomalyResult{
		AnomalyScore: anomalyScore,
		IsAnomaly:    isAnomaly,
		ZScore:       zScore,
	}
}
